### PR TITLE
enable affinity for GC threads and some bug fixes

### PR DESCRIPTION
**Pull request by Maoni0**
@dotnet-bot test this please

### CODE REVIEW

**Comment by Maoni0** [comment]
@dotnet-bot test Ubuntu x64 Checked Build and Test
@dotnet-bot test OSX x64 Checked Build and Test
@dotnet-bot test CentOS7.1 x64 Debug Build and Test
@dotnet-bot test FreeBSD x64 Checked Build

File: src/gc/gc.cpp [hunk 1]
```diff
@@ -5157,29 +5157,27 @@ void set_thread_affinity_mask_for_heap(int heap_number, GCThreadAffinity* affini
             {
                 if (bit_number == heap_number)
                 {
-                    dprintf (3, ("Using processor %d for heap %d\n", proc_number, heap_number));
+                    printf ("Using processor %d for heap %d\n", proc_number, heap_number);
```
Comment (janvorli) [inline src/gc/gc.cpp]: I guess this change from dprintf to printf was made just for your debugging purposes.

Comment (Maoni0) [inline src/gc/gc.cpp]: yeah, this is not ready for review. I just wanted to put it through the non windows builds to make sure it compiles ok. I'll ask for a CR when it's ready.

**Comment by Maoni0** [comment]
@dotnet-bot test Windows_NT x64 Release Priority 1 Build and Test

**Comment by Maoni0** [comment]
@dotnet-bot test OSX x64 Checked Build and Test

**Comment by gkhanna79** [comment]
@wtgodbe Have you seen this failure (in OSX)?

http://dotnet-ci.cloudapp.net/job/dotnet_coreclr/job/master/job/checked_osx_tst_prtest/1304/console

**Comment by wtgodbe** [comment]
I have not - looks like this is CI specific though, based on this line from the console output:

> ERROR: Failed to copy artifacts from dotnet_coreclr/master/checked_windows_nt_bld_prtest with filter: **

It looks like this is unrelated to the problems we were having last week, since those were related to copying artifacts from corefx, whereas this is failing on copying artifacts from a coreclr build. I'm not actually sure what's going on here, considering this change doesn't touch netci.groovy, and it looks like the Ubuntu build & test was able to copy artifacts from that Windows build without any problems.

**Comment by Maoni0** [comment]
something clearly changed, if you look at the CIs from yesterday and the day before, most of the time this job failed. but today it seems to be running much better (still seeing occasional failures but I have not gone and looked whether they are due to the same reason). http://dotnet-ci.cloudapp.net/job/dotnet_coreclr/job/master/job/checked_osx_tst_prtest/

**Comment by wtgodbe** [comment]
That is pretty bizarre - There's a very clear divide that seems to happen right at midnight this morning, where all OSX PRtests before then were failing, and most after are passing. That doesn't seem to correspond (chronologically) to any PRs being merged...

**Comment by AndyAyersMS** [comment]
Maybe not that mysterious: I believe @mmitche was planning to reboot all the OSX hosts since there was an accumulating resource leak or something similar causing the failures.

**Comment by Maoni0** [comment]
@dotnet-bot test CentOS7.1 x64 Debug Build and Test

**Comment by Maoni0** [comment]
@dotnet-bot test CentOS7.1 x64 Debug Build and Test

**Comment by mmitche** [comment]
@Maoni0 Yes, CoreCLR had some kind of resource leak that needed reboots to fix.

**Comment by Maoni0** [comment]
@janvorli  please review

**Comment by Maoni0** [comment]
#3380

File: src/gc/gc.cpp [hunk 2]
```diff
@@ -5157,29 +5157,20 @@ void set_thread_affinity_mask_for_heap(int heap_number, GCThreadAffinity* affini
             {
                 if (bit_number == heap_number)
                 {
-                    dprintf (3, ("Using processor %d for heap %d\n", proc_number, heap_number));
+                    dprintf (3, ("Using processor %d for heap %d", proc_number, heap_number));
```
Comment (janvorli) [inline src/gc/gc.cpp]: Is the newline removal intentional? Other dprintf messages still have it.

File: src/vm/win32threadpool.h [hunk 3]
```diff
@@ -1022,12 +1022,12 @@ class ThreadpoolMgr
         {
             LIMITED_METHOD_CONTRACT;
 
-	    if (CPUGroupInfo::CanEnableGCCPUGroups() && CPUGroupInfo::CanEnableThreadUseAllCpuGroups())
-            return pRecycledListPerProcessor[CPUGroupInfo::CalculateCurrentProcessorNumber()][memType];
-        else
-            // Turns out GetCurrentProcessorNumber can return a value greater than the number of processors reported by
-            // GetSystemInfo, if we're running in WOW64 on a machine with >32 processors.
-        	return pRecycledListPerProcessor[GetCurrentProcessorNumber()%NumberOfProcessors][memType];
+	        if (CPUGroupInfo::CanEnableGCCPUGroups() && CPUGroupInfo::CanEnableThreadUseAllCpuGroups())
```
Comment (janvorli) [inline src/vm/win32threadpool.h]: A nit - the "if" is misaligned by one space. And the same is true for the comments below.

**Comment by janvorli** [comment]
LGTM modulo the nits

File: src/gc/gc.cpp (hunk 2 above)
Comment (Maoni0) [inline src/gc/gc.cpp]: yes it's intentional

File: src/vm/win32threadpool.h (hunk 3 above)
Comment (Maoni0) [inline src/vm/win32threadpool.h]: are you ok if I fix this with my next checkin? I don't want to submit another checkin just for this as it seems very difficult in general to get these checkins through due to one problem or another.

Comment (Maoni0) [inline src/vm/win32threadpool.h]: actually I just checked in the editor and this lines up correctly. it's the previous code that was misaligned.

Comment (janvorli) [inline src/vm/win32threadpool.h]: Ok, feel free to merge it in.

**Pull request merged**